catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/ResultIterable.java

```diff
@@ -77,20 +80,75 @@ public ResultIterable(CatalogFramework catalogFramework, QueryRequest queryReque
      * @param queryRequest  request used to retrieve the results.
      */
     public ResultIterable(QueryFunction queryFunction, QueryRequest queryRequest) {
+        this(queryFunction, queryRequest, 0);
+    }
+
+    /**
+     * Creates an iterable that will call the {@link CatalogFramework} to retrieve the results
+     * that match the {@link QueryRequest} provided.
+     *
+     * @param catalogFramework reference to the {@link CatalogFramework} to call to retrieve the
+     *                         results.
+     * @param queryRequest     request used to retrieve the results.
+     * @param maxResultCount   maximum number of results in total to query for
+     */
+    public ResultIterable(CatalogFramework catalogFramework, QueryRequest queryRequest,
+            int maxResultCount) {
+        notNull(catalogFramework, "CatalogFramework cannot be null");
+        notNull(queryRequest, "Query request cannot be null");
+        Validate.isTrue(maxResultCount >= 0,
+                "Max Results must be zero or positive",
+                maxResultCount);
+
+        this.queryFunction = catalogFramework::query;
+        this.queryRequest = queryRequest;
+        this.maxResultCount = maxResultCount;
+    }
+
+    /**
+     * Creates an iterable that will call a {@link QueryFunction} to retrieve the results
+     * that match the {@link QueryRequest} provided.
+     *
+     * @param queryFunction  reference to the {@link QueryFunction} to call to retrieve the
+     *                       results.
+     * @param queryRequest   request used to retrieve the results.
+     * @param maxResultCount maximum number of results in total to query for
+     */
+    public ResultIterable(QueryFunction queryFunction, QueryRequest queryRequest,
+            int maxResultCount) {
         notNull(queryFunction, "Query function cannot be null");
         notNull(queryRequest, "Query request cannot be null");
+        Validate.isTrue(maxResultCount >= 0,
+                "Max Results must be zero or positive",
+                maxResultCount);
 
         this.queryFunction = queryFunction;
         this.queryRequest = queryRequest;
+        this.maxResultCount = maxResultCount;
     }
 
     @Override
     public Iterator<Result> iterator() {
+        if (maxResultCount > 0) {
+            return limit(new QueryResultIterator(queryFunction, queryRequest), maxResultCount);
+        }
         return new QueryResultIterator(queryFunction, queryRequest);
     }
 
+    /**
```

Comment:
Oh yes I could make it return an iterable instead of an iterator. That would make sense if someone is using it in a for-each. 

Though a counter point, If a client needs a spliterator would they not simply use the spliterator over the results? Wouldn't a spliterator over Lists of results be slightly redundant?

In my eyes it just seems that a large majority of the code is needlessly copied directly. As seen by this implementation there is hardly any real logic needed to produce the functionality that is desired. I'm not sure what we gain by duplicating a large amount of code and keeping it in a separate class. In order to apply the maxResultCount change in it's current state I would essentially need to duplicate my logic in both classes. However in the new condensed version of the ResultListIterator it gets this for almost no effort.

 I know that ResultIterable has already been used extensively. It seems that the most common case is simply for this plain iterator over results. What factored in for the need of creating the List version of this iterable initially? 

In response to the Iterable returning an Iterable, Do we gain much by having the ResultListIterable as a top level class as opposed to just being able to access it via a method (`.listIterator`)-- Especially if it is not frequently used? Forcing to people do have an extra call on the class for a .listIterator won't make me lose much sleep at night, especially with the simplicity it introduces. If the need for it as a top level class increases , we could always split it out and duplicate it again